File: src/server/api/modelParams.ts
export type ModelAlias = 'sonnet' | 'opus' | 'haiku';
export type EffortLevel = 'low' | 'medium' | 'high' | 'xhigh' | 'max';

export interface ModelParams {
  model?: ModelAlias;
  effort?: EffortLevel;
  subagent_model?: ModelAlias;
}

const MODELS: ReadonlySet<string> = new Set(['sonnet', 'opus', 'haiku']);
const EFFORTS: ReadonlySet<string> = new Set([
  'low', 'medium', 'high', 'xhigh', 'max',
]);

export type ValidationResult =
  | { ok: true }
  | { ok: false; message: string };

export function validateModelParams(p: {
  model?: string | null;
  effort?: string | null;
  subagent_model?: string | null;
}): ValidationResult {
  // Treat null and undefined identically as "not provided". This matches how
  // the client may serialize an unset value (either dropped or explicit null).
  const model = p.model ?? undefined;
  const effort = p.effort ?? undefined;
  const subagent = p.subagent_model ?? undefined;

  if (model !== undefined && !MODELS.has(model)) {
    return { ok: false, message: `invalid model: ${model}` };
  }
  if (effort !== undefined && !EFFORTS.has(effort)) {
    return { ok: false, message: `invalid effort: ${effort}` };
  }
  if (subagent !== undefined && !MODELS.has(subagent)) {
    return { ok: false, message: `invalid subagent_model: ${subagent}` };
  }
  if (effort !== undefined && model === 'haiku') {
    return { ok: false, message: 'effort is not supported on haiku' };
  }
  if (effort === 'xhigh' && model !== undefined && model !== 'opus') {
    return { ok: false, message: 'xhigh effort is only supported on opus' };
  }
  return { ok: true };
}

File: server-elixir/lib/fbi/runs/model_params.ex
defmodule FBI.Runs.ModelParams do
  @moduledoc "Validates run model/effort/subagent_model. Mirrors src/server/api/modelParams.ts."

  @models ~w(sonnet opus haiku)
  @efforts ~w(low medium high xhigh max)

  @spec validate(map()) :: :ok | {:error, String.t()}
  def validate(params) when is_map(params) do
    model = Map.get(params, :model) || Map.get(params, "model")
    effort = Map.get(params, :effort) || Map.get(params, "effort")
    subagent = Map.get(params, :subagent_model) || Map.get(params, "subagent_model")

    cond do
      model not in [nil | @models] ->
        {:error, "invalid model: #{model}"}

      effort not in [nil | @efforts] ->
        {:error, "invalid effort: #{effort}"}

      subagent not in [nil | @models] ->
        {:error, "invalid subagent_model: #{subagent}"}

      effort != nil and model == "haiku" ->
        {:error, "effort is not supported on haiku"}

      effort == "xhigh" and model != nil and model != "opus" ->
        {:error, "xhigh effort is only supported on opus"}

      true ->
        :ok
    end
  end
end
